### PR TITLE
doc: Add `near-network` reference level documentation `Message transportation layers`

### PR DESCRIPTION
**Pull request by pmnoxx**
This is part of NetworkDocumentation.md it covers:

-  Message transportation layers

See https://github.com/near/nearcore/issues/5156

### CODE REVIEW

File: docs/network.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
+# 10. Message transportation layers.
+
+This section describes different protocols of sending messages currently used in `Near`
+
+## 10.1 Messages between Actors.
+
+`Near` is build on `Actix`'s `actor` framework. (https://actix.rs/book/actix/sec-2-actor.html)
+Usually each actor runs on its own dedicated thread.
+Only messages implementing `actix::Message`, can be sent using between threads.
+Comment on lines +7 to +9
```
Comment (mm-near) [inline docs/network.md:10]: ??

File: docs/network.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
+# 10. Message transportation layers.
+
+This section describes different protocols of sending messages currently used in `Near`
+
+## 10.1 Messages between Actors.
+
+`Near` is build on `Actix`'s `actor` framework. (https://actix.rs/book/actix/sec-2-actor.html)
+Usually each actor runs on its own dedicated thread.
```
Comment (mm-near) [inline docs/network.md:8]: do we have cases, where we have multiple 'copies' of the actors (therefore running on a few separate threads?)

File: docs/network.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+Only messages implementing `actix::Message`, can be sent using between threads.
```
Comment (mm-near) [inline docs/network.md:9]: I'd also mention that those actors have their own queues - and that the processing of the requests is happening asynchronously.

Comment (mm-near) [reply docs/network.md:9]: (but as these are all 'within' one program - we do not need to worry about backwards compatibility.

File: docs/network.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+impl actix::Message for PeersRequest {
+    type Result = PeerRequestResult;
+}
+pub struct PeerRequestResult {
+    pub peers: Vec<PeerInfo>,
+}
+```
+
+## 10.2 Messages sent through TCP
+Near using `borsh` serialization to exchange messages between nodes (See https://borsh.io/).
```
Comment (mm-near) [inline docs/network.md:31]: Near *is* using

File: docs/network.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+Only messages implementing `BorshSerialize`, `BorshDeserialize` can be sent.
+
```
Comment (mm-near) [inline docs/network.md:33]: (we also use borsh for storage)

File: docs/network.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+Comment on lines +7 to +9
+
+ bowenwang1996 4 days ago Member
+
+We should not leak implementation details into the spec
+
+	Reply…
+Resolve conversation
+
+On example of such message is `PeersRequest` :
```
Comment (mm-near) [inline docs/network.md:19]: I'd just mention, that people should look for 'impl actix::Message for XX' to see which messages can be passed such way

File: docs/network.md [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+}
+```
+
+## 10.2 Messages sent through TCP
+Near using `borsh` serialization to exchange messages between nodes (See https://borsh.io/).
+Only messages implementing `BorshSerialize`, `BorshDeserialize` can be sent.
+
+Here is an example of on such message:
+```rust
+#[derive(BorshSerialize, BorshDeserialize)]
```
Comment (mm-near) [inline docs/network.md:36]: this example is probably not needed (just saying that everything that has a 'derived BorshSerialize/Deserialize' can be sent over the network / written to storage)

File: docs/network.md [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+pub struct SyncData {
+    pub(crate) edges: Vec<Edge>,
+    pub(crate) accounts: Vec<AnnounceAccount>,
+}
+```
+
+## 10.3 Messages sent/received through `chain/jsonrpc`
+Near runs a json rest server. (See `actix_web::HttpServer`).
```
Comment (mm-near) [inline docs/network.md:44]: s/rest/REST

File: docs/network.md [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+All messages sent and received must implement `serde::Serialize` and `serde::Deserialize`.
+
```
Comment (mm-near) [inline docs/network.md:46]: and I'd mention that with such messages, we should keep backwards compatibility - until we are sure that customer updated/adjusted their code.

File: docs/network.md (hunk 7 above)
Comment (mm-near) [reply docs/network.md:36]: which means that we have to be VERY careful if we're changing any of those - as it is basically a protocol change.

**Review by mm-near — DISMISSED** [review]

File: docs/network.md (hunk 2 above)
Comment (pmnoxx) [reply docs/network.md:131]: Yes. some of them have one thread per instance.

**Review by mm-near — APPROVED** [review]

**Pull request merged**